### PR TITLE
Add MergeLimitWithDistinct rule

### PR DESCRIPTION
**Pull request by martint**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeLimitWithDistinct.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        PlanNode input = lookup.resolve(parent.getSource());
+        if (!(input instanceof AggregationNode)) {
+            return Optional.empty();
+        }
+
+        AggregationNode child = (AggregationNode) input;
+
+        if (!child.getAggregations().isEmpty() ||
+                child.getOutputSymbols().size() != child.getGroupingKeys().size() ||
+                !child.getOutputSymbols().containsAll(child.getGroupingKeys())) {
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeLimitWithDistinct.java]: is this optimizer effected by grouping sets?

Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeLimitWithDistinct.java]: It only works if the aggregation is doing a "distinct", which means there's a single grouping set containing all columns.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeLimitWithDistinct.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import com.facebook.presto.sql.planner.iterative.Lookup;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.AggregationNode;
+import com.facebook.presto.sql.planner.plan.DistinctLimitNode;
+import com.facebook.presto.sql.planner.plan.LimitNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+
+import java.util.Optional;
+
+public class MergeLimitWithDistinct
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeLimitWithDistinct.java]: it looks like you forgot to add `@Deprecated` to legacy optimizer rule

Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeLimitWithDistinct.java]: I still haven't ported all the "limit pushdown" optimizations, that's why I haven't done it.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeLimitWithDistinct.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        LimitNode parent = (LimitNode) node;
+
+        PlanNode input = lookup.resolve(parent.getSource());
+        if (!(input instanceof AggregationNode)) {
+            return Optional.empty();
+        }
+
+        AggregationNode child = (AggregationNode) input;
+
+        if (!child.getAggregations().isEmpty() ||
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeLimitWithDistinct.java]: extract `isDistinctNode(AggregationNode)`.
Or maybe make it inctance method of `AggregationNode`

**Pull request closed without merging**